src/PowerShellEditorServices/Services/DebugAdapter/DebugService.cs

```diff
@@ -792,55 +805,95 @@ private bool AddToAutoVariables(PSObject psvariable, string scope)
             // process.  Without the intermediate variable Get-PSCallStack inexplicably returns
             // an array of strings containing the formatted output of the CallStackFrame list.
-            var callStackVarName = $"$global:{PsesGlobalVariableNamePrefix}CallStack";
-            psCommand.AddScript($"{callStackVarName} = Get-PSCallStack; {callStackVarName}");
+            var callStackVarName = $"$GLOBAL:{PsesGlobalVariableNamePrefix}CallStack";
 
-            var results = await _executionService.ExecutePSCommandAsync<PSObject>(psCommand, CancellationToken.None).ConfigureAwait(false);
+            var getPSCallStack = $"Get-PSCallStack | % {{ [void]{callStackVarName}.add(@($PSItem,$PSItem.GetFrameVariables())) }}";
+
+            // If we're attached to a remote runspace, we need to serialize the callstack prior to transport
+            // because the default depth is too shallow
+            var isOnRemoteMachine = _psesHost.CurrentRunspace.IsOnRemoteMachine;
+            var returnSerializedIfOnRemoteMachine = isOnRemoteMachine
+                ? $"[Management.Automation.PSSerializer]::Serialize({callStackVarName}, {serializationDepth})"
+                : callStackVarName;
+
+            // We have to deal with a shallow serialization depth with ExecutePSCommandAsync as well, hence the serializer to get full var information
+            // TODO: Don't use serializer for local runspaces, or implement a way to specify depth ExecutePSCommandAsync
+            string getPSCallStackScript = $"[Collections.ArrayList]{callStackVarName}=@();{getPSCallStack};{returnSerializedIfOnRemoteMachine}";
+            psCommand.AddScript(getPSCallStackScript);
 
-            var callStackFrames = results.ToArray();
 
-            this.stackFrameDetails = new StackFrameDetails[callStackFrames.Length];
+            // PSObject is used here instead of the specific type because we get deserialized objects from remote sessions and want a common interface
+            var results = await _executionService.ExecutePSCommandAsync<PSObject>(psCommand, CancellationToken.None).ConfigureAwait(false);
+
+            IEnumerable callStack = isOnRemoteMachine
+                ? (PSSerializer.Deserialize(results[0].BaseObject as string) as PSObject).BaseObject as IList
+                : results;
 
-            for (int i = 0; i < callStackFrames.Length; i++)
+            List<StackFrameDetails> stackFrameDetailList = new();
+            foreach (var callStackFrameItem in callStack)
             {
-                VariableContainerDetails autoVariables =
-                    new VariableContainerDetails(
-                        this.nextVariableId++,
-                        VariableContainerDetails.AutoVariablesName);
+                var callStackFrameComponents = (callStackFrameItem as PSObject).BaseObject as IList;
+                var callStackFrame = callStackFrameComponents[0] as PSObject;
```

Comment:
Consider a null check somewhere around here.  Maybe `callStackFrameComponents?[0] as PSObject` and a way to break out after this.  If it's believed to be impossible consider using `Debug.Assert` to make transitioning to nullable reference types easier later on

Reply:
If we have a callstack then it is going to have these components, it would have to be an API change otherwise, so getting a null exception might be a good "indicator" there that something changed, all I'd be doing is throwing a more specific exception at best maybe. Again may be my C# noobness talking, but switching to nullable the compilers going to highlight this line anyways.

Reply:
Maybe an explicit cast would be better then. Explicit cast will throw, `as` means "if it's this type gimme it, otherwise null".  Well it won't throw on `null` but still, intent is a little bit more clear